Adopt TypeScript for core package (#1877)
MarkBind/markbind: docs/devGuide/migratingToTypeScript.md

3. **Rename the files** from having `.js` extensions to `.ts` extensions.

4. **Stage the changes** (the renamed files, `.gitignore`, and `.eslintignore`) for commit. Ensure the renamed files are regarded by Git as `renamed` and not `added`.

5. **Commit the staged changes** with the message `Rename <file-or-folder> to TypeScript`.
   <br>
   For example: `Rename core/src/html to TypeScript`.

### Step 3: "Adapt" Commit

1. **Enable automatic compilation of TypeScript files** on save/change.

   You can read more about this on [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.

2. **Change the CommonJS module import/export syntax** to TypeScript's equivalent (use ES6 only if necessary).

   A common error in newly renamed files comes from the way modules are imported/exported. The syntax for those can be converted according to the table below:

   CommonJS                | TypeScript equivalent    | ES6
   ------------------------|--------------------------|-----------------------
   `const ... = require()` | `import ... = require()` | `import ... from ...`
   `module.exports = {}`   | `export = {}`            | `export ...`

>    `module.exports = ...`  | `export = ...`           | `export default ...`

Reviewer: > Do not use the ES6 syntax if not necessary, as it compiles into something more than just the CommonJS syntax when compiled, and might lead to unexpected behaviours when another `.js` file imports the compiled file, to the point of failing established tests in some cases.

This particular line may not be equivalent (commonJs dosen't have concept of `default`), and might be the sole cause of your issues with above quote.

Try inserting the following in the playground [here](https://www.typescriptlang.org/play?module=1#code/PTAEAEGcAsHsHcCiBbAlgFwFAgs2ATAVwBsBTALlAGNZk8A7AK0mzHHtkQCcvYuXUyAA590oAN6gAbgENihUgHkAZgAVUoAL6hlvZKABEAOmA16kdDPrpIBgNyZMpAB4iuYsxdDp4sdaABeaTkFFX8AKlAAJgcnV1FQfFJlGRIxHz9UByA)

```
// @showEmit
// @module: commonjs
// @noErrors
import { valueOfPi } from "./constants";

export const twoPi = valueOfPi * 2;

export default twoPi;
```

Reviewer: I think one solution may be to advise against using `export default` in our **intra** (external libraries should be fine) exports / imports altogether (we've been mostly using named exports too which should make this easier)

Author: Agree with `export default`. I have already disabled `prefer-default-export` in ESLint, maybe we can strengthen it by enabling `no-default-export`. Does external libraries here mean the ones that will be consumed by other packages (e.g. `cli`)? If so, we still have to follow how `cli/index.js` imports `core/index.js` to see our options right now.

As a note, I was just about to write that I remembered I couldn't make ES6-style named exports (`export ...`) work in `CyclicReferenceError.ts` as it will always fail the tests, suggesting that it's not always possible to use ES6-style. But, as I investigated further to write this comment, I found that the root cause is that there is a file (`Context.js`) that imports the `CyclicReferenceError` class as a *default* export. Because `export default` is out of the question, therefore the only way for `CyclicReferenceError.ts` to fit in with other files is to use the TypeScript-style default export (`export = ...`).

I can update the text so that we can start from TypeScript-style, then move to ES6 *if possible*, but do not use `export default`.

Reviewer: > Does external libraries here mean the ones that will be consumed by other packages (e.g. cli)?

Nope, just regular external libraries (e.g. `simple-git`).

> I can update the text so that we can start from TypeScript-style, then move to ES6 if possible, but do not use export default.

Sounds good 👍
(at least for now, while we're still migrating, default exports won't play so well with js)

> I found that the root cause is that there is a file (Context.js) that imports the CyclicReferenceError class as a default export.

> default export.

Rather, the `module.exports` object. (commonjs has no "default" export, by default 😅)

Just to be clear, this would actually work, but rather clunky-ly:

```
var x = require('ts-module').default;
```

Once we have almost all files converted we can possibly revisit if `export default` would be a good fit (if really needed).

   Start with the TypeScript equivalent syntax, as it is compiled to be exactly the same as the CommonJS syntax, making it the safest choice when adapting the file. You can read more about this syntax in the [official documentation](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require).

   There are some limitations to the TypeScript equivalent syntax, such as:
   - There can only be one `export =` in a file

>    - You cannot reference what is not (unlike `module.exports`)

Reviewer: 👀

   - A module exported with `export =` must be imported with the TypeScript equivalent syntax (for `.ts` files) or the CommonJS syntax (for `.js` files).

   Do not use the ES6 syntax if not necessary, as it compiles into something more than just the CommonJS syntax when compiled, and might lead to unexpected behaviours when another `.js` file imports the compiled file, to the point of failing established tests in some cases.

   However, changing to the ES6 syntax might be necessary, e.g. you cannot use the TypeScript equivalent syntax if you want to import/export types or interfaces. When doing so, make sure the files pass the tests first before proceeding.

> 3. **Adapt the files** to fully using TypeScript.

Reviewer: ```suggestion
3. **Adapt the files** fully to TypeScript.
```

>    The errors from TypeScript and `typescript-eslint` can guide you on what to fix. Some tips on adapting to TypeScript:

Reviewer: ```suggestion
   The errors from TypeScript and `typescript-eslint` can guide you on what to fix. Here are some tips:
```

When we rename files, we have to keep the Git similarity index between the old file and the new file to be above a certain threshold (50%) to keep the history, or else it will be regarded as a different file with its own history.

Combining the two steps into one - that is, adapting files immediately after renaming - will lower the similarity index, therefore making it possible for the files, especially small ones or those that need a lot of changes when adapting, to hit below the similarity threshold and lose their history.

The solution to this is to make the rename and adapt commits separate, in order for Git to recognize first that the file has changed into `.ts`, and so the changes are compared not against the old `.js` file, but to the renamed `.ts` file instead.

### Why do we need a merge commit

Even if the migration developer has kept the history intact through the separate "Rename" and "Adapt" commits, this is only intact at their working branch. At the end of the day, the totality of the changes in the working branch is compared against the target branch, in which the same consideration would apply.

> If we do the usual squash commit, the two separate commits are combined into one, and so the problem with losing history occurs again, but now in the context of the target branch. Again, the solution is to keep those two commits separate, even in the target branch, and therefore what we must merge the PR with a merge commit to achieve this.

Reviewer: 👀 
Dosen't a merge PR still squash everything into one commit in the target branch? (hence the need for `rename-threshold`)

But preserving the history in the working branch, from which git can _possibly_ infer the rename.
(In contrast, squash removes the information from the working branch completely)

Author: Hmm yeah, now that I think about it, I felt like I was only basing it like a fast-forward merge rather than an actual merge commit haha. Will update into it.

<span id="title" class="d-none">{{ title }}</span>

<frontmatter>
  title: "{{ title }}"
  layout: devGuide.md
  pageNav: default
</frontmatter>

# {{ title }}

<div class="lead">

In order to improve the quality and security of our backend code, we are introducing TypeScript to our codebase to leverage the type-checking system. This page outlines the steps of migrating existing files into TypeScript.
</div>

## Migration Steps

The TypeScript migration process is a little different than normal development work, as maintainers would not follow the normal procedure of creating a squash commit. Instead, they will do a **normal merge commit**, and therefore migration developers will need to structure the commits in a particular way.

We have decided on structuring the commits as two commits: a "Rename" commit and an "Adapt" commit. If you need further context on the approach, feel free to read the [Explanation Notes](#explanation-notes) section below.

### Step 1: Preparation

1. **If running, stop automatic compilation on change/save**. More on this at [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.

>    <br>

Author: Will remove this

   If the dependency has its typings:

   - Bundled alongside their implementation, you don't need to do anything else.

   - In a `@types/<lib-name>` module, install it as a development dependency for the `core` package by navigating to `packages/core` and running `npm i -D @types/<lib-name>`.

     After installing all necessary typings, delete the generated `packages/core/package-lock.json`, navigate back to the root directory, and run `npm run setup`.

     <box type="info">

     It's recommended to try to match the `@types/<lib-name>` version with the associated module. In general, try to match the major and minor version as closely as possible.
     </box>

3. **Stash the changes** (`packages/core/package.json` and root `package-lock.json`).
   <br>
   This will be added back during the "Adapt" commit later.

### Step 2: "Rename" Commit

> 1. **Add the paths to the `.js` files to `.gitignore` and `.eslintignore`**.

Reviewer 2: ```suggestion
1. **Add the paths of the `.js` files to `.gitignore` and `.eslintignore`**.
```

### Step 3: "Adapt" Commit

1. **Start automatic compilation on change/save**. More on this at [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.

2. **Change the CommonJS module import/export syntax** to TypeScript's equivalent syntax (use ES6 syntax only if possible and/or necessary).

   A common error in newly renamed files comes from the way modules are imported/exported. The syntax for those can be converted according to the table below:

   Case           | CommonJS                   | TypeScript equivalent                                            | ES6
   ---------------|----------------------------|------------------------------------------------------------------|-----------------------
   Default import | `const x = require(y)`     | `import x = require(y)`                                          | `import x from y`
   Named import   | `const { x } = require(y)` | N/A, workaround:<br>`import z = require(y)`<br>`const { x } = z` | `import { x } from y`
   Named export   | `module.exports = { x }`   | `export = { x }`                                                 | `export { x }`

>    Default export | `module.exports = x`       | `export = x`                                                     | `export default x`

Reviewer: referring https://github.com/MarkBind/markbind/pull/1877#discussion_r851785006

Just one nit as such while you're still touching things up -- let's make it clear the regular `import x = require(y)` / `export = x` isn't default import export. (Maybe just give a warning, and link the typescript docs / playground)

E.g. `import x from y` actually compiles to this:

![ts](https://user-images.githubusercontent.com/3306138/163726482-119d6bfe-c446-4d2a-8dd7-0754e6fcb324.png)

---

Remember to push fixups as well! https://github.com/MarkBind/markbind/pull/1877#pullrequestreview-943841714

Author: Apologies if my definitions might lead to some confusion, but to clarify, the "default" import/export I'm mentioning here is rather more to the concept of importing/exporting a single (non-object) thing, such a constant, a function, a class, etc. It is why I associate `const x = require(y)`, `import x = require(y)`, and `import x from y` together.

I have also noted in the docs that, as exampled in the image above, the ES6 syntax will compile into something more than just CommonJS syntax, and thus proceed with ES6 only if possible. I guess what I'm trying to say is that while they are doing the same job at high-level (import/export a single thing), they are not identical at low-level.

I will update the docs to clarify this more.

Reviewer: maybe:
- "module.exports object"
- "Entire export"? (still seems ambiguous)
- "`export` object"

Author: I think there is no agreeable term for the concept. Most just simply say one is to "export one thing" and the other to "export many things". There are developer-written articles (like [this](https://www.freecodecamp.org/news/node-module-exports-explained-with-javascript-export-function-examples/) and [this](https://www.sitepoint.com/understanding-module-exports-exports-node-js/)) kind of use the same definition as I wrote earlier, "export one thing" = "default export" in CommonJS context.

Author: I just updated the docs to clarify this, and since it is becoming too long I put the syntax reference on a separate section to be referred to from the migration steps. I opt for the simpler "export one thing" vs "export many things" to describe the different cases.

Reviewer: > I think there is no agreeable term for the concept. Most just simply say one is to "export one thing" and the other to "export many things". There are developer-written articles (like [this](https://www.freecodecamp.org/news/node-module-exports-explained-with-javascript-export-function-examples/) and [this](https://www.sitepoint.com/understanding-module-exports-exports-node-js/)) kind of use the same definition as I wrote earlier, "export one thing" = "default export" in CommonJS context.

No, do refer to a canonical, or at least peer reviewed source. https://nodejs.org/api/modules.html, https://stackoverflow.com/questions/21117160/what-is-export-default-in-javascript, https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6

There is no concept of "default import/export" in commonJs, this is a very common misconception that the first article you linked has clearly made as well. The second article refers to the vague concept of a "default value", not "default import/export" (which is a technical term formalised in es6 proposals).

This is important as its a very common misconception, one that I made when I started learning about this as well, and as you mentioned has very real implementation implications. We have contributors who aren't as familiar with module systems coming in; Developer documentation above all else should aim to be technically accurate and unambiguous.

The canonical term here would simply be `module.exports` or `exports` object (https://nodejs.org/api/modules.html).

Default exports, are in fact, closer to the concept of named exports in the context of commonJS. https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6

> I opt for the simpler "export one thing" vs "export many things" to describe the different cases.

Sounds better, and good work here 👍 

Some minor suggestions dropped below

>    Similar to the process of refactoring a function, you might have to **adjust how these files are imported by other TypeScript (`.ts`) files**. If you are changing it to the TypeScript equivalent syntax, then it must be imported with so as well. The same goes for the ES6 syntax. 

Reviewer 2: Not sure if there is a typo here?
```suggestion
   Similar to the process of refactoring a function, you might have to **adjust how these files are imported by other TypeScript (`.ts`) files**. If you are changing it to the TypeScript equivalent syntax, then it must be imported the same way as well. The same goes for the ES6 syntax. 
```

   The ES6 syntax compiles into something more than just the CommonJS syntax when compiled, and might lead to unexpected behaviours when another `.js` file imports the compiled file, to the point of failing tests in some cases. You can try changing to the ES6 syntax, but only proceed if there are no failing tests. The TypeScript equivalent syntax is enough for the initial adaptation.

   However, in some cases, changing to the ES6 syntax might be necessary. For instance, you cannot use the TypeScript equivalent syntax if you want to import/export TypeScript specific constructs, such as types, interfaces, and enums. Again, when doing so, make sure that all tests pass.

   <box type="info">

   We will perform a mass change into ES6 syntax after all files in core have been adapted into TypeScript.
   </box>

3. **Adapt the files** to fully use TypeScript.

>    The errors from TypeScript and `typescript-eslint` can guide you on what to fix. Only fix what is necessary and do not modify any code functionality. Avoid using `any` as best as you can.

Reviewer: ```suggestion
   The errors from TypeScript and `typescript-eslint` can guide you on what to fix. Only fix what is necessary and be careful with accidentally modifying any code functionality. Avoid using `any` as best as you can.
```

>    If you happen to encounter type errors related to using names from an internal dependency that is still in JavaScript (`.js`), you can try to infer simple types in your file first as a "stand-in" of the more robust types when that internal dependency is eventually migrated.

Reviewer: ```suggestion
   If you happen to encounter type errors related to using names from another MarkBind import that is still in JavaScript (`.js`), you can try to infer simple types in your file first as a "stand-in" of the more robust types when that internal dependency is eventually migrated.
```

> related to using names

👀 what's this ("names")?

Author: It's my attempt to generalize referring to constants, properties, members, etc. The case that I saw that , was that a file in migration is trying to access a static class member of another file that is still in `.js`. We have to make do for now inferring simple types like `Record<string, any>` for example while we're waiting for the other `.js` file to be migrated.

>    On the flip side of the above situation, as you now have developed a robust type for your own file, you might have to **adjust how other TypeScript (`.ts`) files type names referenced from your files**. Some type declarations might need to be made stricter, or can be removed if it's not necessary anymore, and so on. This will ensure that the files will be incrementally stricter. 

Reviewer: ```suggestion
   On the flip side of the above situation, once you have developed a robust type for your own file, **adjust how other TypeScript (`.ts`) files type names referenced from your files** as applicable. Some type declarations might need to be made stricter, or can be removed if it's not necessary anymore, and so on. This will ensure that the files will be incrementally stricter. 
```

4. **Make sure everything is in order** by running `npm run test` in the root directory.

>    The `core` package test directly tests `.ts` files (powered by `ts-jest`), and the `cli` package test uses the compiled files from `core`. Therefore, passing the `cli` tests means that it is very likely the compiled files work as expected.

Reviewer 2: ```suggestion
   The `core` package test can directly run on the `.ts` files (powered by `ts-jest`), and the `cli` package test uses the compiled files from `core`. Therefore, passing the `cli` tests means that it is very likely the compiled files work as expected.
```

5. **Stage the adapted files and the typings from Step 1** for commit. Verify the changes before committing.

6. **Commit the changes** with the message `Adapt <file-or-folder> to TypeScript`.
   <br>
   For example, `Adapt core/src/html to TypeScript`.

You are now ready to create a pull request for the changes to the repository.

## Example of Migrated Works

You can see these pull requests to observe the finished migration works:

- [#1877: Adopt TypeScript for core package](https://github.com/MarkBind/markbind/pull/1877)

## Explanation Notes

> ### Why do we need separate "Rename" and "Adapt" commits

Reviewer 2: ```suggestion
### Why do we need to separate "Rename" and "Adapt" commits
```

We want to keep the file history intact throughout this migration, meaning that we should still be able to see the change history of the `.ts` files even from the time when they were still in `.js`.

> When we rename files, we have to keep the Git similarity index between the old file and the new file to be above a certain threshold (50%) to keep the history, or else it will be regarded as a different file with its own history.

Reviewer 2: ```suggestion
When we rename files, we have to keep the Git similarity index between the old file and the new file above a certain threshold (50%) to keep the history, or else it will be regarded as a different file with its own history.
```

1. **Start automatic compilation on change/save**. More on this at [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.

2. **Change the CommonJS module import/export syntax** to TypeScript's equivalent syntax (use ES6 syntax only if possible).

>    A common error in newly renamed files comes from the way modules are imported/exported. Those statements can be converted according to the [Import/Export Syntax Reference](#importexport-syntax-reference) section below. The sequence is as follows:

Reviewer: > ~sequence is~ steps are as follows

(for clarity that these are steps that should be followed, not a sequence of X)

>    1. Start by changing the import/export statements into TypeScript equivalent syntax. Be careful with the imports as you have to match them with the export syntax that the modules use, be it TypeScript equivalent or ES6.

Reviewer: > Be careful with ~the~ imports, as you have

>    2. Check whether it is possible to change the export use ES6 syntax. As mentioned in the syntax reference, ES6 should not be used if the module only exports a single thing. Change to ES6 syntax only if possible, otherwise keep with TypeScript equivalent syntax.

Reviewer: > export use ES6 syntax

> export to use ES6 syntax

>    3. You might have to **adjust how the files are imported by other TypeScript (`.ts`) files**. Before, you would see that some If you are changing the export to the TypeScript equivalent syntax, then it must be imported the same way as well. The same goes for the ES6 syntax.

Reviewer: > You might have to **adjust how the files are imported by other TypeScript (`.ts`) files**. Before, you would see that some If you are changing the export to the TypeScript equivalent syntax, then it must be imported the same way as well. The same goes for the ES6 syntax.

> You might have to **adjust how the files are imported by other TypeScript (`.ts`) files**. ~Before~Previously, you would see that ~some~ if you are changing the export to the TypeScript equivalent syntax, then it must be imported the same way as well. The same goes for the ES6 syntax.

Author: Some excess statement there 😅 removing the `Before, you would see that some` part instead.

3. **Adapt the files** fully to TypeScript.

   The errors from TypeScript and `typescript-eslint` can guide you on what to fix. Only fix what is necessary and be careful with accidentally modifying any code functionality. Avoid using `any` as best as you can.

>    If you happen to encounter type errors related to using names from another MarkBind import that is still in JavaScript (`.js`), you can try to infer simple types in your file first as a "stand-in" of the more robust types when that internal dependency is eventually migrated.

Reviewer: > related to using names from another MarkBind import

> related to using names (of constants, properties, members, etc.) from another MarkBind import that is still written in JavaScript

- [#1877: Adopt TypeScript for core package](https://github.com/MarkBind/markbind/pull/1877)

## Import/Export Syntax Reference

Here is the reference on changing the import/export statements from CommonJS syntax into those that TypeScript support: TypeScript equivalent syntax, and ES6 syntax.

The TypeScript equivalent syntax is designed to compile to the exact same code as CommonJS. You can read more about this syntax in the [official documentation](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require).

> The ES6 syntax is the newer JavaScript syntax that TypeScript recommends, but in certain cases it does not compile to the exact same code as CommonJS.

Reviewer: > ~The~ ES6 syntax is ~the~a newer ... exact same code as CommonJS. We simplify the concepts here for illustration, refer to the StackOverflow post [here](https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6) for more details!

> In the tables below, a "thing" is defined as a JavaScript construct that can be exported, such as constants, functions, classes, etc. In the ES6 syntax, a "thing" can also be TypeScript constructs like types, interfaces, enums, etc.

Reviewer: > In ~the~ ES6 syntax

!!**Exports**!!

| What to Export              | CommonJS                         | TypeScript equivalent   | ES6                   |
|-----------------------------|----------------------------------|-------------------------|-----------------------|

> | One thing only              | `module.exports = x`             | `export = x`            | `export default x`*   |

Reviewer: 👀  would footnotes look better / ok? (on the "*" and "^"'s)

could also try bolding it if not to make it stand out from normal text a little bit more

Author: I'm not sure on the footnotes. I feel like these are important info that readers should understand, and putting it in our footnotes component hides it from view unless the reader intentionally hovers to it.

My intention for the symbols is to contextualize which syntax the info is referring to. In a way, it's like footnotes, but I just wanted to make the content more prominent than actual footnotes.

Reviewer: formatting nit to make the asterixes appear more prominent:

`<span style="font-weight:bold;">&nbsp;*</span>`

could also stick to `*` / one symbol to make it consistent.

Author: :+1: I'll take the `^` symbol as it's more visible than `*`

| Object of one/more things   | `module.exports = { x, y, z }`   | `export = { x, y, z }`  | `export { x, y, z }`  |

> *: For compatibility, **do not use** `export default` during migration. It is compiled differently, and does not play well with dependants that are still in `.js` (unless some modifications are made in the `.js` files, which would be quite clunky).

Reviewer: > For compatibility, **do not use** `export default` during migration. It is compiled differently, and does not play well with dependants that are still in `.js`, unless some clunky modifications are made in the `.js` files (refer to the StackOverflow post previously [linked](https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6)!).

Even if the migration developer has kept the history intact through the separate "Rename" and "Adapt" commits, this is only intact at their working branch. At the end of the day, the totality of the changes in the working branch is compared against the target branch, in which the same consideration would apply.

If we do the usual squash commit, the changes from the two commits are combined into a new commit and only that commit will be pushed into the target branch. The original two commits are omitted, therefore the history of the working branch that we have tried to keep intact is stripped away.

> The normal merge commit also creates another commit (the merge commit), but the merge process interleaves the commits from the working branch to the target branch, therefore retaining the changes history.

Reviewer: > change~s~ history
